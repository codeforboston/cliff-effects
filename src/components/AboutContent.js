import React from 'react';
import { Header, Message } from 'semantic-ui-react';
import { ExternalLink } from '../forms/formHelpers';

// CUSTOM COMPONENTS

/** Todo simplify the process of creating content for the 'About' page */

const AboutContent = function ({ snippets }) {
  return (
    <div>

      <Header
        as='h1'
        style={{ fontSize: '4em' }}>
        { snippets.aboutPageHeader_v1 }
      </Header>

      <Header as='h3'>{ snippets.whatForHeader_v1 }</Header>
      <Message>{ snippets.whatForImportantNote_v1 }</Message>
      <p>{ snippets.whatFor_v2 }</p>

      <Header as='h3'>{ snippets.whyHeader_v1 }</Header>
      <p>{ snippets.why1_v1 }</p>
      <p>{ snippets.why2_v1 }</p>

      <ul>
        <li>
          <ExternalLink href="https://www.youtube.com/watch?v=BveX_rID4_E">
            { snippets.videoLinkText_v1 }
          </ExternalLink>
        </li>

        <li>
          <ExternalLink href="http://www.nccp.org/projects/files/NCCP_CO_presentation07.pdf">
            { snippets.quantLinkText_v1 }
          </ExternalLink>
        </li>

        <li>
          <ExternalLink href="https://www.umb.edu/editor_uploads/images/centers_institutes/center_social_policy/The_Road_to_the_Cliff_Edge_08.16.17.pdf">
            { snippets.benefitsLinkText_v1 }
          </ExternalLink>
        </li>
      </ul>

      <Header as='h3'>{ snippets.howToUseHeader_v1 }</Header>
      <p>{ snippets.howToUse_v1 }</p>
      <Message>{ snippets.howToUseNote_v1 }</Message>

      <Header as='h3'>{ snippets.whoMadeThisHeader_v1 }</Header>
      <p>{ snippets.whoMadeThis1_v1 }</p>
      <p>{ snippets.whoMadeThis2_v1 }</p>
      <p>{ snippets.whoMadeThis3_v1 }</p>

    </div>
  );

};  // End AboutContent(<>)


export { AboutContent };
