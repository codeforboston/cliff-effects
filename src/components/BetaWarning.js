import React from 'react';
import { Menu, Responsive } from 'semantic-ui-react';

const BetaWarning = function (props) {
  return (
    <Menu.Item>
      <Responsive
        as='strong'
        {...Responsive.onlyTablet}>
        This tool is a prototype.
      </Responsive>
      <Responsive
        as='strong'
        {...Responsive.onlyComputer}>
        This tool is a prototype and should not be used to make financial decisions.
      </Responsive>
    </Menu.Item>
  );
};  // End BetaWarning(<>)


export { BetaWarning };
