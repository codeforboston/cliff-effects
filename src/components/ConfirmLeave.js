import React from 'react';

/**
 * Ask the user for confirmation to leave the page.
 * @extends React.Component
 *
 * @param props {object}
 * @param props.isBlocking {boolean} - Whether the component should block
 * @param props.message {string} - The message the browser shows, maybe
 */
class ConfirmLeave extends React.Component {
  static defaultProps = { isBlocking: true };

  confirm = (event) => {
    if (!this.props.isBlocking) {
      return; // do not block unload
    } 
    return (event.returnValue = this.props.message || '');
  };

  componentDidMount() {
    window.addEventListener('beforeunload', this.confirm);
  }

  componentWillUnmount() {
    window.removeEventListener('beforeunload', this.confirm);
  }

  render() {
    return null;
  }
}

export default ConfirmLeave;
