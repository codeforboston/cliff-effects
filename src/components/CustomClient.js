import React from 'react';
import { Divider, Form, Message } from 'semantic-ui-react';

/**
 * Load previous session from object.
 * @callback loadClient
 * @param {object} clientContainer - Object containing previous session as `client`.
 */

/**
 * Form which loads previous session from JSON.
 * 
 * @param {object} props
 * @param {boolean} props.mayLoadCustomClient - Whether form should be visible
 * @param {loadClient} props.loadClient
 * 
 * @extends React.Component
 */
class CustomClient extends React.Component {
  state = {
    client: null,
    error:  null,
    json:   '',
  };

  submit = (event) => {
    const { client } = this.state;
    event.preventDefault();
    if (client === null) {
      return;
    }

    this.setState({
      client: null,
      error:  null,
      json:   '',
    });
    this.props.loadClient({ client: client });
  };

  handleChange = (_event, inputProps) => {
    const { value } = inputProps;
    try {
      const newClient = JSON.parse(value);
      this.setState({
        client: newClient,
        error:  null,
        json:   value,
      });
    } catch (error) {
      this.setState({
        client: null,
        error:  error,
        json:   value,
      });
    }
  };

  render() {
    if (!this.props.mayLoadCustomClient) {
      return null;
    }
    const { client, error, json } = this.state;

    return (
      <Form
        error={ error !== null }
        onSubmit={ this.submit }>
        <Form.Field>
          <label>Client JSON</label>
          <Form.Input
            type={ 'text' }
            value={ json }
            onChange={ this.handleChange } />
        </Form.Field>
        <Message 
          error
          header={ 'JSON Parse Failed!' }
          content={ error && error.message } />
        <Form.Button
          type={ 'submit' }
          disabled={ client === null }>
          Import Data
        </Form.Button>
        <Divider />
      </Form>
    );
  }
}

export { CustomClient };
