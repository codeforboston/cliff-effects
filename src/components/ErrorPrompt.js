import React from 'react';

/**
 * Prompt for latest uncaught error.
 * @extends React.Component
 *
 * @param props {object}
 * @param props.callback {onLeaveCallback}
 * @param props.message {string}
 * @param props.prompt {prompt}
 */
class ErrorPrompt extends React.Component {
  handleError = ({ error }) => {
    const { callback, prompt, ...promptProps } = this.props;
    prompt(callback, promptProps);
  };

  componentDidMount() {
    window.addEventListener('error', this.handleError);
  }

  componentWillUnmount() {
    window.removeEventListener('error', this.handleError);
  }

  render() {
    return null;
  }
}

export default ErrorPrompt;
