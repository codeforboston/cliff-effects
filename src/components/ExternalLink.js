import React from 'react';
/**
 * Link that opens new tab
 */
const ExternalLink = function ({ href, children, ...otherProps }) {
  return (
    <a
      href   = { href }
      target = { `_blank` }
      { ...otherProps }>
      {children}
    </a>
  );
};

export { ExternalLink };
