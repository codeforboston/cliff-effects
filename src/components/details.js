import React, { Component } from 'react';
import {
  Icon,
  Transition,
  Message,
} from 'semantic-ui-react';


/** Really it's a top component that has a question mark
 *    next to it followed by a bottom one that starts
 *    as hidden, but is revealed when the question mark
 *    is clicked.
 *
 * @param {object} props
 * @param {object} props.children Must have exactly two
 *    children. The first is always visible. The second
 *    is revealed when the icon is clicked and then hidden
 *    when the icon is clicked again.
 */
class HeadingWithDetail extends Component {
  state = { showDetails: false };

  toggleDetails = () => {
    this.setState((prevState) => {
      return { showDetails: !prevState.showDetails };
    });
  };

  render () {
    var { showDetails } = this.state;

    var children  = this.props.children;
    if (!Array.isArray(children) || children.length !== 2) {
      console.warn(`<HeadingWithDetail> must have exactly two children`);
      return null;
    }

    var top       = children[ 0 ] || children,
        details   = children[ 1 ] || null,
        // Has a green question mark icon by default
        iconName  = `question circle outline`,
        iconColor = `teal`;

    // Show a red 'x' instead if the detail is visible right now
    if (showDetails) {
      iconName  = `remove circle`;
      iconColor = `red`;
    }

    return (
      <div className = { `heading-with-detail` }>
        <div className={ `has-details` }>{ top }</div>
        <Icon
          name      = { iconName }
          color     = { iconColor }
          className = { `details-icon` }
          onClick   = { this.toggleDetails } />
        <Transition
          visible   = { showDetails }
          animation = { `slide down` }>
          <Message>{ details }</Message>
        </Transition>
      </div>
    );
  };
};  // End <HeadingWithDetail>


/** @todo Have rows that can have details. Maybe other things too. */
class RowWithDetail extends Component {};  // <End RowWithDetail>


export {
  HeadingWithDetail,
  RowWithDetail,
};
