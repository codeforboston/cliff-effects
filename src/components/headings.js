// REACT COMPONENTS
import React from 'react';
import { Header } from 'semantic-ui-react';


/** h1 styles available for form (and other) text.
*
* @function
* @param {object} props
* @property {string} props.subheading - For when some
*     kind of clarifier is needed.
* @property {object} props.children - Ususally a string
*     to be shown as the heading text.
*
* @returns Component
*/
const ContentH1 = function ({ subheading, children }) {

  if (!children) {
    return null;
  }

  return (
    <div className={ 'text-h1' } >
      <div /> {/** div here to make sure header margin doesn\'t collapse */}
      <Header
        as='h3'
        style={{ display: 'inline-block' }}>
        { children }
      </Header>
      <ContentSubH1>{ subheading }</ContentSubH1>
      <br />
    </div>
  );

};  // End ContentH1{} Component


/** A clearer way than a ternary operator to have a possible
* subheader and separate styling.
*
* @function
* @param {object} props
* @property {object} props.children - Contents of this element
*
* @returns Component
*/
const ContentSubH1 = function ({ children }) {

  if (!children) {
    return null;
  }

  return (
    <div
      className = { 'text-sub-h1' }
      style={{ display: 'block', textAlign: 'left' }}>
      { children }
    </div>
  );

};  // End ContentSubH1{} Component


/* To discuss: Should form-specific headings be in a different file? */


/** Weekly/Monthly/Yearly headings combined for the
*     top of columns that need those time intervals.
*
* @function
* @param {object} props
* @property {object} props.type - Very fragile. Used to
*     create the heading for the column for labels of the
*     inputs - labels like 'Earned Income'. This was created
*     for income and expenses columns to try to cue the user
*     about the info they were putting in. The current uses
*     are to create the headings 'Income Type' and 'Expenses
*     Type', so the user will have an extra hint about what
*     information they should be giving at that moment.
*
* @returns Component
*/
const IntervalColumnHeadings = function ({ type }) {

  let columnTitle = type.toLowerCase().replace(/\b[a-z]/g, (letter) => {
        return letter.toUpperCase();
      }) + ' Type',
      styles      = { fontSize: '14px' };

  return (
    <div style={{ display: 'inline-block' }}>
      <ColumnHeading
        type={ type }
        colName='weekly'
        style={ styles }>Weekly
      </ColumnHeading>
      <ColumnHeading
        type={ type }
        colName='monthly'
        style={ styles }>Monthly
      </ColumnHeading>
      <ColumnHeading
        type={ type }
        colName='yearly'
        style={ styles }>Yearly
      </ColumnHeading>
      <ColumnHeading
        type={ type }
        colName={ type }
        style={ styles }
        columnTitle={ columnTitle }>{columnTitle}
      </ColumnHeading>
    </div>
  );

};  // End <IntervalColumnHeadings>


/** Style for text at the tops of columns, like
*     cashflow or household columns.
*
* @function
* @param {object} props
* @property {string} props.type - Used for element classes
* @property {string} props.colName - Also for the element classes
* @property {object} props.style - Style override
* @property {object} props.children - Usually the heading text
*
* @returns Component
*/
const ColumnHeading = function ({ type, colName, style, children }) {
  let classes = type + '-column cashflow-column header ' + colName;
  return (
    <Header
      as='h4'
      className={ classes }
      style={ style }
      color='teal'>
      { children }
    </Header>
  );
};  // End <ColumnHeading>


export {
  ContentH1,
  ContentSubH1,
  ColumnHeading,
  IntervalColumnHeadings,
};
