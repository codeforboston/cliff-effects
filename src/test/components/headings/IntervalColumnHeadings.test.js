import React from 'react';
import { mount } from 'enzyme';

import { IntervalColumnHeadings } from '../../../components/headings';

describe('<IntervalColumnHeadings>', () => {
  it('renders with uppercase type', () => {
    expect(mount(<IntervalColumnHeadings type="ALL CAPS" />)).toMatchSnapshot();
  });

  it('renders with lowercase type', () => {
    expect(mount(<IntervalColumnHeadings type="all lowercase" />)).toMatchSnapshot();
  });

  it('renders with alphanumeric type', () => {
    expect(mount(<IntervalColumnHeadings type="1WITH 2NUMBERS" />)).toMatchSnapshot();
  });
});
