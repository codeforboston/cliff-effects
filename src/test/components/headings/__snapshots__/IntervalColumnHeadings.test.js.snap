// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IntervalColumnHeadings> renders with alphanumeric type 1`] = `
<IntervalColumnHeadings
  type="1WITH 2NUMBERS"
>
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <ColumnHeading
      colName="weekly"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="1WITH 2NUMBERS"
    >
      <Header
        as="h4"
        className="1WITH 2NUMBERS-column cashflow-column header weekly"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header 1WITH 2NUMBERS-column cashflow-column header weekly"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Weekly
        </h4>
      </Header>
    </ColumnHeading>
    <ColumnHeading
      colName="monthly"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="1WITH 2NUMBERS"
    >
      <Header
        as="h4"
        className="1WITH 2NUMBERS-column cashflow-column header monthly"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header 1WITH 2NUMBERS-column cashflow-column header monthly"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Monthly
        </h4>
      </Header>
    </ColumnHeading>
    <ColumnHeading
      colName="yearly"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="1WITH 2NUMBERS"
    >
      <Header
        as="h4"
        className="1WITH 2NUMBERS-column cashflow-column header yearly"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header 1WITH 2NUMBERS-column cashflow-column header yearly"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Yearly
        </h4>
      </Header>
    </ColumnHeading>
    <ColumnHeading
      colName="1WITH 2NUMBERS"
      columnTitle="1with 2numbers Type"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="1WITH 2NUMBERS"
    >
      <Header
        as="h4"
        className="1WITH 2NUMBERS-column cashflow-column header 1WITH 2NUMBERS"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header 1WITH 2NUMBERS-column cashflow-column header 1WITH 2NUMBERS"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          1with 2numbers Type
        </h4>
      </Header>
    </ColumnHeading>
  </div>
</IntervalColumnHeadings>
`;

exports[`<IntervalColumnHeadings> renders with lowercase type 1`] = `
<IntervalColumnHeadings
  type="all lowercase"
>
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <ColumnHeading
      colName="weekly"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="all lowercase"
    >
      <Header
        as="h4"
        className="all lowercase-column cashflow-column header weekly"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header all lowercase-column cashflow-column header weekly"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Weekly
        </h4>
      </Header>
    </ColumnHeading>
    <ColumnHeading
      colName="monthly"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="all lowercase"
    >
      <Header
        as="h4"
        className="all lowercase-column cashflow-column header monthly"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header all lowercase-column cashflow-column header monthly"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Monthly
        </h4>
      </Header>
    </ColumnHeading>
    <ColumnHeading
      colName="yearly"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="all lowercase"
    >
      <Header
        as="h4"
        className="all lowercase-column cashflow-column header yearly"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header all lowercase-column cashflow-column header yearly"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Yearly
        </h4>
      </Header>
    </ColumnHeading>
    <ColumnHeading
      colName="all lowercase"
      columnTitle="All Lowercase Type"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="all lowercase"
    >
      <Header
        as="h4"
        className="all lowercase-column cashflow-column header all lowercase"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header all lowercase-column cashflow-column header all lowercase"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          All Lowercase Type
        </h4>
      </Header>
    </ColumnHeading>
  </div>
</IntervalColumnHeadings>
`;

exports[`<IntervalColumnHeadings> renders with uppercase type 1`] = `
<IntervalColumnHeadings
  type="ALL CAPS"
>
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <ColumnHeading
      colName="weekly"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="ALL CAPS"
    >
      <Header
        as="h4"
        className="ALL CAPS-column cashflow-column header weekly"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header ALL CAPS-column cashflow-column header weekly"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Weekly
        </h4>
      </Header>
    </ColumnHeading>
    <ColumnHeading
      colName="monthly"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="ALL CAPS"
    >
      <Header
        as="h4"
        className="ALL CAPS-column cashflow-column header monthly"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header ALL CAPS-column cashflow-column header monthly"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Monthly
        </h4>
      </Header>
    </ColumnHeading>
    <ColumnHeading
      colName="yearly"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="ALL CAPS"
    >
      <Header
        as="h4"
        className="ALL CAPS-column cashflow-column header yearly"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header ALL CAPS-column cashflow-column header yearly"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          Yearly
        </h4>
      </Header>
    </ColumnHeading>
    <ColumnHeading
      colName="ALL CAPS"
      columnTitle="All Caps Type"
      style={
        Object {
          "fontSize": "14px",
        }
      }
      type="ALL CAPS"
    >
      <Header
        as="h4"
        className="ALL CAPS-column cashflow-column header ALL CAPS"
        color="teal"
        style={
          Object {
            "fontSize": "14px",
          }
        }
      >
        <h4
          className="ui teal header ALL CAPS-column cashflow-column header ALL CAPS"
          style={
            Object {
              "fontSize": "14px",
            }
          }
        >
          All Caps Type
        </h4>
      </Header>
    </ColumnHeading>
  </div>
</IntervalColumnHeadings>
`;
